dvc/remote/local/__init__.py

```diff
@@ -573,3 +573,55 @@ def _get_unpacked_dir_names(self, checksums):
             if self.is_dir_checksum(c):
                 unpacked.add(c + self.UNPACKED_DIR_SUFFIX)
         return unpacked
+
+    def _needs_checkout(self, path_info, checksum, force):
+        # NOTE: In case if path_info is already cached, and cache type is
+        # 'copy' we would like to avoid relinking.
+        # `symlink/hardlink/reflink` is cheap, but if we need `copy` links,
+        # then to save on copying we should only relink if we already have
+        # `hardlink` or a `symlink`, we don't care about reflinks, because
+        # they are indistinguishable from copy anyway.
+
+        if (
+            not force
+            and not self.changed(path_info, {self.PARAM_CHECKSUM: checksum})
+            and not System.is_hardlink(path_info)
+            and not System.is_symlink(path_info)
+            and self._is_cache_copy(path_info)
+        ):
+            if self.protected:
+                self.protect(path_info)
+            else:
+                # NOTE: we can unprotect, because `hardlink/symlink` check
+                # has been performed before, so no chance of copying,
+                # only chmod-ing.
+                self.unprotect(path_info)
+
+            msg = "File '{}' didn't change"
+            logger.debug(msg.format(str(path_info)))
+            return False
+        return True
+
+    def _is_cache_copy(self, path_info):
+        # NOTE: path_info required to make test reliable, when cache is on
+        # different fs than path_info
+
+        if self.cache_types[0] == "copy":
+            return True
+
+        link_test_dir = self.path_info / "tmp_test"
+        self.makedirs(link_test_dir)
+
+        file1 = link_test_dir / uuid()
+        file2 = path_info.parent / uuid()
+
+        with open(fspath_py35(file1), "wb") as fobj:
+            fobj.write(bytes(1))
+
+        try:
+            self.link(file1, file2)
```

Comment:
You are trying to link _from_ workspace to _cache_, which is backwards.

Reply:
Oh, you are linking from workspace to workspace, because both test_dir and file2 are in the workspace. This is wrong.

Reply:
That is not true, main difference is that file one is under `self.path_info` while file2 is under `path_info.parent`. Those are different entities.

Reply:
@pared those are different, but they are both on the same fs most likely and you need to test cache dir fs vs workspace fs, because they might be the same or might be different, on which the link type is going to depend.

Reply:
Ah, wait, I did misread it. self.path_info is effectively a cache dir, so this is fine, indeed. Sorry for the noise :slightly_smiling_face: